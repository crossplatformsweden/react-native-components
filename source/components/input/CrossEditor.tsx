import React from 'react';
import { TextInput, TextInputProps } from 'react-native-paper';
import { TextInputMask, TextInputMaskProps } from 'react-native-masked-text';

/**
 * Properties for the {@link CrossEditor} component. Supports masking through {@link maskProps}. This allows getting the masked value.
 *
 * The `onChangeText` event will isntead export the raw value.
 *
 * Inherits react-native-paper {@link https://callstack.github.io/react-native-paper/text-input.html TextInputProps}
 */
export interface ICrossEditorProps extends TextInputProps {
  /**
   * Text for floating label
   * https://callstack.github.io/react-native-paper/text-input.html
   */
  label: string;
  /**
   * `True` to use error styling
   * https://callstack.github.io/react-native-paper/text-input.html
   */
  error?: boolean | undefined;
  /**
   * Visual appearence. Default value is 'outlined'
   *
   * https://callstack.github.io/react-native-paper/text-input.html
   */
  mode?: 'flat' | 'outlined' | undefined;
  /**
   * Allows masking the input.
   *
   * https://github.com/benhurott/react-native-masked-text
   *
   * @example
   * ```
   <CrossEditor
    label='Mobile'
    maskProps={{ type: 'cel-phone' }} />```
   */
  maskProps?: TextInputMaskProps | undefined;
}

let maskInput: any | null = null;

/**
 * A text input supporting mask through {@see ICrossEditorProps.maskProps}.
 *
 * When using mask the raw value will be exported in the `onChangeText` event, not the masked value.
 *
 * Default {@see ICrossEditorProps.mode mode} is 'outlined'.
 *
 * Props are {@see ICrossEditorProps}
 *
 *
 * @example
 * ```
 <CrossEditor
  label='Mobile'
  maskProps={{ type: 'cel-phone' }} />
 ```
 */
export class CrossEditor extends React.Component<ICrossEditorProps> {
  constructor(props: ICrossEditorProps) {
    super(props);
  }

  render() {
    const paperProps = {
      ...this.props,
      mode: this.props.mode || ('outlined' as 'flat' | 'outlined'),
    } as TextInputProps;
    // console.log('** CrossEditor paper props = ', paperProps);

    if (this.props.maskProps && this.props.maskProps.type) {
      const maskedProps = {
        ...this.props,
        ...this.props.maskProps,
      } as TextInputMaskProps;
      // console.log('** CrossEditor masked props = ', maskedProps);

      // console.log('*** CrossEditor: render TextInputMask');

      return (
        <TextInputMask
          {...maskedProps}
          ref={(comp) => (maskInput = comp)}
          // @ts-ignore - bad map
          customTextInput={TextInput}
          customTextInputProps={paperProps}
          onChangeText={() => {
            // console.log('**** CrossEditor: onChangeText');
            if (!maskInput || !maskInput.getRawValue) {
              // console.log('TextInputMask missing getRawValue', maskInput);
              return;
            }

            // @ts-ignore - bad map
            const rawValue = maskInput.getRawValue();
            // console.log('CrossEditor rawValue = ', rawValue || 'nothing');

            if (maskedProps && maskedProps.onChangeText) {
              maskedProps.onChangeText(rawValue);
              return;
            }

            if (paperProps.onChangeText) {
              paperProps.onChangeText(rawValue);
              return;
            }

            // console.log('***** CrossEditor: no listener *****');
          }}
        />
      );
    }

    // console.log('** CrossEditor: render paper TextInput **');
    return <TextInput {...paperProps} mode={paperProps.mode} />;
  }
}

export default CrossEditor;
