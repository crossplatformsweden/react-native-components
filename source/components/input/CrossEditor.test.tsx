// https://itnext.io/testing-react-16-3-components-with-react-test-renderer-without-enzyme-d9c65d689e88
import React from 'react';
import { ICrossEditorProps, CrossEditor } from './CrossEditor';
import TestRenderer from 'react-test-renderer';
import { TextInputMaskProps } from 'react-native-masked-text';

jest.unmock('react-native');
jest.unmock('./CrossEditor');

/**
   * Test component rendering. Properties of children might be tested by importing their type:
   *```
      const btn = wrapper.root.findByType(Button); // or findByProps({ mode: 'text' });
      expect(btn.props.title).toBe('child');
      ```
   */
describe('components', () => {
  describe('<CrossEditor />', () => {
    describe('`ICrossEditorProps` interface', () => {
      const propsMock: ICrossEditorProps = { label: 'Test' };
      it('Should match snapshot', () => {
        expect(propsMock).toMatchSnapshot();
      });

      it('Should have `maskProps` property that is undefined', () => {
        expect(propsMock.maskProps).toBeUndefined();
      });
    });

    describe('With `label` property provided', () => {
      it('Component should render', () => {
        const wrapper = TestRenderer.create(
          <CrossEditor
            label={'Test'}
            onChangeText={jest.fn()}
            value={'Textvalue'}
          />
        );
        expect(wrapper.toJSON()).toMatchSnapshot();
      });

      it('Should have paper <TextInput /> with mode "outlined"', () => {
        const wrapper = TestRenderer.create(
          <CrossEditor
            label={'Test'}
            onChangeText={jest.fn()}
            value={'Textvalue'}
          />
        );
        const child = wrapper.root.findByProps({ mode: 'outlined' });
        expect(child.type).toEqual('TextInput');
      });

      it('TextInputProps onChangeText should be called', () => {
        let gotText: string | undefined = undefined;
        const onCalled = (val: string) => (gotText = val);

        const textVal = '111222';
        const maskProps: TextInputMaskProps = { type: 'zip-code' };

        const wrapper = TestRenderer.create(
          <CrossEditor
            onChangeText={onCalled}
            label={'Test'}
            maskProps={maskProps}
            value={'default value'}
          />
        );

        const child = wrapper.root.findByProps(maskProps);
        child.instance.props.onChangeText(textVal);

        // There will not be a value test time
        expect(gotText).toEqual('');
      });

      it('maskProps onChangeText should be called', () => {
        let gotText: string | undefined = undefined;
        const onCalled = (val: string) => (gotText = val);

        const textVal = '111222';
        const maskProps: TextInputMaskProps = {
          type: 'zip-code',
          onChangeText: onCalled,
        };

        const wrapper = TestRenderer.create(
          <CrossEditor
            label={'Test'}
            maskProps={maskProps}
            value={'default value'}
          />
        );

        const child = wrapper.root.findByProps({
          type: 'zip-code',
        });
        child.instance.props.onChangeText(textVal);

        // There will not be a value test time
        expect(gotText).toEqual('');
      });
    });
  });
});
