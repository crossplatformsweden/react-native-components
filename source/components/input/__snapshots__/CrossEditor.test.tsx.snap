// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components <CrossEditor /> \`ICrossEditorProps\` interface Should match snapshot 1`] = `
Object {
  "label": "Test",
}
`;

exports[`components <CrossEditor /> With \`label\` property provided Component should render 1`] = `
<TextInput
  label="Test"
  mode="outlined"
  onChangeText={[MockFunction]}
  value="Textvalue"
/>
`;
